### PR TITLE
Hyp Smearing Merge

### PR DESCRIPTION
**Pull request by goracle**
I followed the instructions on https://latticeqcd.github.io/SIMULATeQCD/02_contributions/git.html .  All the tests passed (prior to the merge into main), although the main branch seems to have a different script to run the tests (which I did not run).  I'm not sure how git push is supposed to create a pull request, so I am creating the pull request as I usually do (via a fork).

Hyp smearing here is only 4D, and it has been tested (by me) on only one 1 GPU.

### CODE REVIEW

File: src/gauge/constructs/fat7LinkConstructs.h
Comment (dbollweg) [inline src/gauge/constructs/fat7LinkConstructs.h:28]: I think it would be good to put the hyp-smearing specific changes into a separate threeLinkStaple and call it hypThreeLinkStaple_first_level or something consistent with the other hyp functions in order to keep it separated from the HISQ functions.

Comment (goracle) [reply src/gauge/constructs/fat7LinkConstructs.h:28]: I agree, that makes sense.  I did not spend very long thinking about naming conventions for this port, so I am generally fine with whatever changes need to be made in that regard.

**Review by lukas-mazur — CHANGES_REQUESTED** [review]

File: src/base/math/gsu3.h
Comment (goracle) [reply src/base/math/gsu3.h:19]: I don't recall adding this line.  Perhaps it came from an earlier version of the code?  I know of know reason this should be here.  Feel free to delete.

File: src/base/communication/communicationBase_mpi.cpp
Comment (goracle) [reply src/base/communication/communicationBase_mpi.cpp:680]: I mentioned this in an e-mail to Rasmus, so I'll copy my response.

"Without this change, I get stack smashing errors on the GPU machine at
MSU.  I traced it with gdb to this line.  I did confirm today that the
crash still happens with my latest commit, making the change back to
getCart_comm().  The crash occurs regardless of how many mpi processes
I run. 
"

So in summary I don't know why it fails for getCart_comm(), but since it is just opening a file perhaps it is ok to use the global MPI communicator.

File: src/base/math/gcomplex.h
Comment (goracle) [reply src/base/math/gcomplex.h:148]: source for that?  I was mostly using this to debug, although I don't think comparison operators introduce round-off.  I do think it is true that underflows used to be truncated to zero, but this hasn't been true for awhile, and certainly that would happen to the inputs prior to a comparison operator like ==.

Without this line, comparisons to zero did fail silently, so == 0.0 seems better than nothing.  It is true that relative comparison of floats makes no sense for comparisons to zero, even if the norm multiplies the relative difference epsilon.  However, here we have an operator named == acting on floating point numbers, but under the hood it for some reason does not behave how == does for c++ floating point numbers.  I appreciate the desire to write something better than what c++ has for this operator, but it seems confusing from my perspective to ask for == and get back a relative comparison that also drops support for comparison to zero and also fails silently for zero comparison.

In the interest of having zero comparison not fail silently and also not have a=b=0 be the only near zero point supported, though (I mean it should not fail for numbers near zero but succeed for zero), let me offer the following:

we have the current comparison test in cmp_rel as

    if (abs(a-b) / abs(a+b) < rel && abs(a-b) < prec) {

Certainly the main failure mode, without specifying rel and prec, is a,b near zero.  Otherwise, this seems sensible, although without knowing what cmp_rel is supposed to do exactly I can't say for sure if it is correct.

I think the general solution, though, should be something like https://stackoverflow.com/a/32334103 for float comparison.  The current solution we are discussing (with comparison == 0.0) is qualitatively the first graph under "Graphs please?"  I think the approach given in the link is (broadly speaking) what I'd support, and I'd be interested to know if people can think of any improvements (besides generic type support).  This link also shows us how to build cmp_rel, in that cmp_rel should be a function like nearly_equal(a,b), but with an assert that causes cmp_rel to fail on comparisons near zero.

If the above were implemented, the point where relative comparison breaks down becomes machine/implementation/type dependent.  However, use of various types like float already subjects us to this dependence.  If we are near these limits, we should expect a lot of problems, not just in our comparison operator.  This should be an indication that the user is using the wrong floatT, rather than a problem with the function/operator.

File: src/base/communication/communicationBase_mpi.cpp
Comment (lukas-mazur) [reply src/base/communication/communicationBase_mpi.cpp:680]: That sounds like a special problem with the MSU machine. Could you please check whether this is also an issue on another machine (e.g. Bielefeld Cluster) ?

File: src/base/math/gcomplex.h
Comment (luhuhis) [reply src/base/math/gcomplex.h:148]: I just want to mention that also issue #29 exists, which is probably related

File: src/base/communication/communicationBase_mpi.cpp
Comment (goracle) [reply src/base/communication/communicationBase_mpi.cpp:680]: Sure.  Does the Bielefeld cluster have a GPU node?  I do have an account on the Bielefeld cluster, but I don't currently have access to any compute resources there, I think.  I have just been using my account there to back up data to tape.

File: src/base/math/gcomplex.h
Comment (RasmusNL) [reply src/base/math/gcomplex.h:148]: The current version

if (abs(a-b) / abs(a+b) < rel && abs(a-b) < prec) return true

Will indeed give problems for a+b = 0

The link Dan provided will solve this. 

I think the minimalist way to solve this problem is to change the code into

if ( 2.0*abs(a-b)  < rel*(abs(a)+abs(b))+1e-37 && abs(a-b) < prec) return true

If you don't like the small constant added (chosen to be around 10 times smaller that the smallest float number) one can replace 

rel*(abs(a)+abs(b))+1e-37 with  max(rel*(abs(a)+abs(b)), smallest float number)

Last thing. I think the if statement is not needed, should just be

return (2.0*abs(a-b)  < rel*(abs(a)+abs(b))+1e-37 && abs(a-b) < prec)

This does assume that a and b are smaller than half the maximum float number. A check for this can be added if wanted.

File: src/base/communication/communicationBase_mpi.cpp
Comment (clarkedavida) [reply src/base/communication/communicationBase_mpi.cpp:680]: Yes, if you have an account you can use our GPUs. It looks like you have a Redmine account, which means you can learn how to use our GPU cluster here: https://rmp.physik.uni-bielefeld.de/projects/cluster

On our Volta machine, which is equipped with V100 GPUs, you can get a 4-GPU interactive node with
`run --partition=volta_devel --nodes=1 --ntasks=4 --gres-flags=enforce-binding --gres=gpu:4 --time=00:30:00 --gpus=4 --pty $SHELL`

**Comment by goracle** [comment]
I made all the changes requested (please let me know if I missed one), and the hyp test still passes.  I assume the other tests also still pass, but I did not run them.  Cheers.

File: src/testing/main_HypSmearingTest.cu
Comment (lukas-mazur) [inline src/testing/main_HypSmearingTest.cu:47]: Please replace this readconf line with a dynamically generated config from "GaugeUpdate::set_gauge_to_reference()" and remove lat.sample.l6666_nersc from the branch

**Review by lukas-mazur — CHANGES_REQUESTED** [review]

File: src/testing/main_HypSmearingTest.cu
Comment (goracle) [reply src/testing/main_HypSmearingTest.cu:47]: Hi Lucas,

I based the Hyp smearing code on the Hisq smearing code, which still has:
gauge_in.readconf_nersc("../test_conf/gauge12750");
in main_HisqSmearingTest.cpp
I would request clarification on if this code is also going to be modified to use the dynamic scheme or not.  I think there should be one policy for both tests, in principle.

Also, I validated the hyp smearing code based on the lattice included.  It seems this is a request that I output the dynamically generated lattice, smear it with the Milc code, and then check that the output is the same.  If this is not necessary, than it seems we are saying that the barrier to the merge is not the correctness of the smearing.  If that is the case, can you yourself be the one to make the change after the merge?  I'm not much interested in the semantics of the test (or which lattice is used).    

Is the dynamic lattice the same from run to run (what if the RNG generator used changes, or the implementation changes how it handles the zero seed)?  The way the test works now is string comparison vs. the outputs of Milc code.  I don't currently have another way to validate it.  If you do have a dynamic validation strategy, though, I would be interested to know what it is and if it can be added as an additional (or substitute) check.  

Cheers,

Comment (goracle) [reply src/testing/main_HypSmearingTest.cu:47]: Hi Lukas,

Confirming the RNG does not depend on implementation (i.e. if I compile it on my laptop it will give the same output as on a KNL) addresses one of my concerns in my above comment.  The other concern I had is that main_HisqSmearingTest.cpp is not using a dynamic lattice.  If this is the policy, then you should make the change there too.  It is slightly beyond the scope of this discussion, but I had a look again at main_HisqSmearingTest.cpp.  The test used is a unitarity test, so it should not be lattice dependent.  However, main_HisqSmearingTest.cpp uses a reference lattice + unitarity test (which is lattice independent), and I don't know how this reference lattice was generated.  If, as you indicate, the policy involved is to accept the correctness of the test as is, then it should be straightforward to generate this reference lattice dynamically.  

My other concern (as stated) is that I'm overall not interested in what lattice is used, so it does not make much sense for me to be further involved in making this change.  You are welcome to pull my changes and make this change afterwards; it is clear that it is not critical in the physics sense whether the change is made before or after (if it were, then main_HisqSmearingTest would've already been changed).  

I can confirm that I did the test with the Milc code as you describe.

Hopefully this resolves this issue; I'm uninterested in repeating myself.

Comment (lukas-mazur) [reply src/testing/main_HypSmearingTest.cu:47]: Dear Dan,

I agree with you that main_HisqSmearingTest could also be updated. This is not surprising, since this test is rather old and the policy that we should use dynamic lattices is rather new. 
An issue to replace the usage of static lattices with dynamic lattices in all tests already exists. 
And someone should definitely take care of that at some point.
However, the fact that main_HisqSmearingTest is not using a dynamic lattice, does not justify that all new contributions are allowed to use configurations from disk.

I really appreciate that you want to make a contribution to the code and I am always happy to help if you face any problems in doing so. 
However, I will not finalize this branch for you.  
If you don't want to do that, you will probably need to find someone else to do it for you. 
Otherwise, we cannot merge this branch. These are the rules.

Regards,
Lukas

**Comment by goracle** [comment]
Ok, please do let me know if you find a problem with the correctness, memory usage, or speed of the code as it is.  Otherwise, it seems like whether or not what I've written is merged hinges on some issues that do not and should not interest me.  Peace.

**Comment by lukas-mazur** [comment]
Fair enough. I guess since you made it clear that your are not interested in finalizing this branch I can close this pull request here.

**Pull request closed without merging**
